projects/com.oracle.truffle.llvm.parser.factories/src/com/oracle/truffle/llvm/parser/factories/LLVMRootNodeFactory.java

```diff
@@ -137,4 +107,88 @@ private static LLVMAddress getArgsAsStringArray(Object[] args) {
         return stringArgs;
     }
 
+    private static String[] getenv() {
+        return System.getenv().entrySet().stream().map((e) -> e.getKey() + "=" + e.getValue()).toArray(String[]::new);
+    }
+
+    private static LLVMAddress putCString(LLVMAddress address, String string) {
+        LLVMAddress ptr = address;
+        for (byte b : string.getBytes()) { // automatic charset conversion
+            LLVMMemory.putI8(ptr, b);
+            ptr = ptr.increment(1);
+        }
+        LLVMMemory.putI8(ptr, (byte) 0);
+        return ptr.increment(1);
+    }
+
+    private static LLVMAddress getArgs(Object[] mainArgs) {
```

Comment:
This function has to be modeled with Sulong nodes.
Use a malloc node to allocate memory, use the free node to free it.
Use the pointer arithmetic nodes to compute address values. 
Use memory access nodes to load/store data to memory.